bigtable/metrics_monitoring_exporter.go

```diff
@@ -0,0 +1,367 @@
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"sync"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	"cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
+	"go.opentelemetry.io/otel/attribute"
+	otelmetric "go.opentelemetry.io/otel/sdk/metric"
+	otelmetricdata "go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"google.golang.org/api/option"
+	"google.golang.org/genproto/googleapis/api/distribution"
+	googlemetricpb "google.golang.org/genproto/googleapis/api/metric"
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	meterName            = "bigtable.googleapis.com/internal/client/"
+	bigtableResourceType = "bigtable_client_raw"
+
+	// The number of timeserieses to send to GCM in a single request. This
+	// is a hard limit in the GCM API, so we never want to exceed 200.
+	sendBatchSize = 200
+)
+
+var (
+	monitoredResLabelsSet = map[string]bool{
+		monitoredResLabelKeyProject:  true,
+		monitoredResLabelKeyInstance: true,
+		monitoredResLabelKeyCluster:  true,
+		monitoredResLabelKeyTable:    true,
+		monitoredResLabelKeyZone:     true,
+	}
+
+	errShutdown = fmt.Errorf("exporter is shutdown")
+)
+
+type errUnexpectedAggregationKind struct {
+	kind string
+}
+
+func (e errUnexpectedAggregationKind) Error() string {
+	return fmt.Sprintf("the metric kind is unexpected: %v", e.kind)
+}
+
+// monitoringExporter is the implementation of OpenTelemetry metric exporter for
+// Google Cloud Monitoring.
+// Default exporter for built-in metrics
+type monitoringExporter struct {
+	shutdown     chan struct{}
+	client       *monitoring.MetricClient
+	shutdownOnce sync.Once
+	projectID    string
+}
+
+func newMonitoringExporter(ctx context.Context, project string, opts ...option.ClientOption) (*monitoringExporter, error) {
+	client, err := monitoring.NewMetricClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &monitoringExporter{
+		client:    client,
+		shutdown:  make(chan struct{}),
+		projectID: project,
+	}, nil
+}
+
+// ForceFlush does nothing, the exporter holds no state.
+func (e *monitoringExporter) ForceFlush(ctx context.Context) error { return ctx.Err() }
+
+// Shutdown shuts down the client connections.
+func (e *monitoringExporter) Shutdown(ctx context.Context) error {
+	err := errShutdown
+	e.shutdownOnce.Do(func() {
+		close(e.shutdown)
+		err = errors.Join(ctx.Err(), e.client.Close())
+	})
+	return err
+}
+
+// Export exports OpenTelemetry Metrics to Google Cloud Monitoring.
+func (me *monitoringExporter) Export(ctx context.Context, rm *otelmetricdata.ResourceMetrics) error {
+	select {
+	case <-me.shutdown:
+		return errShutdown
+	default:
+	}
+
+	return me.exportTimeSeries(ctx, rm)
+}
+
+// Temporality returns the Temporality to use for an instrument kind.
+func (me *monitoringExporter) Temporality(ik otelmetric.InstrumentKind) otelmetricdata.Temporality {
+	return otelmetric.DefaultTemporalitySelector(ik)
```

Comment:
[Default is cumulative](https://github.com/open-telemetry/opentelemetry-go/blob/sdk/metric/v1.27.0/sdk/metric/reader.go#L131) but updated the code to directly use cumulative since default could change.